### PR TITLE
fast-reject: add depth check

### PR DESCRIPTION
**Pull request by lcnr**
slightly modified version of #133524

I tried a few alternatives:
- simply bail after recursion for a certain amount of times, looking at the number of steps taken while compiling different crates we get the following results[^1]. As we need a fairly high limit here using one felt somewhat arbitrary, it also had a slightly larger performance impact than the final approach

typenum
```rust
1098842 counts
(  1)   670511 (61.0%, 61.0%): dropping after 1
(  2)   358785 (32.7%, 93.7%): dropping after 0
(  3)    25191 ( 2.3%, 96.0%): dropping after 2
(  4)    10912 ( 1.0%, 97.0%): dropping after 4
(  5)     6461 ( 0.6%, 97.5%): dropping after 3
(  6)     5239 ( 0.5%, 98.0%): dropping after 5
(  7)     2528 ( 0.2%, 98.3%): dropping after 8
(  8)     2188 ( 0.2%, 98.5%): dropping after 1094
(  9)     2097 ( 0.2%, 98.6%): dropping after 6
( 10)     1179 ( 0.1%, 98.7%): dropping after 34
( 11)     1148 ( 0.1%, 98.9%): dropping after 7
( 12)      822 ( 0.1%, 98.9%): dropping after 10
```
bitmaps
```rust
533346 counts
(  1)   526166 (98.7%, 98.7%): dropping after 1
(  2)     4562 ( 0.9%, 99.5%): dropping after 0
(  3)     2072 ( 0.4%, 99.9%): dropping after 1024
(  4)      305 ( 0.1%,100.0%): dropping after 2
(  5)      106 ( 0.0%,100.0%): dropping after 4
(  6)       30 ( 0.0%,100.0%): dropping after 8
(  7)       18 ( 0.0%,100.0%): dropping after 3
(  8)       17 ( 0.0%,100.0%): dropping after 44
(  9)       15 ( 0.0%,100.0%): dropping after 168
( 10)        8 ( 0.0%,100.0%): dropping after 14
( 11)        7 ( 0.0%,100.0%): dropping after 13
( 12)        7 ( 0.0%,100.0%): dropping after 24
```
stage 2 compiler is mostly trivial, but has a few cases where we get >5000
```rust
12987156 counts
(  1)  9280476 (71.5%, 71.5%): dropping after 0
(  2)  2277841 (17.5%, 89.0%): dropping after 1
(  3)   724888 ( 5.6%, 94.6%): dropping after 2
(  4)   204005 ( 1.6%, 96.2%): dropping after 4
(  5)   146537 ( 1.1%, 97.3%): dropping after 3
(  6)    64287 ( 0.5%, 97.8%): dropping after 5
(  7)    43938 ( 0.3%, 98.1%): dropping after 6
(  8)    43758 ( 0.3%, 98.4%): dropping after 8
(  9)    27220 ( 0.2%, 98.7%): dropping after 7
( 10)    17374 ( 0.1%, 98.8%): dropping after 9
( 11)    16015 ( 0.1%, 98.9%): dropping after 10
( 12)    12855 ( 0.1%, 99.0%): dropping after 12
( 13)    10494 ( 0.1%, 99.1%): dropping after 11
( 14)     7553 ( 0.1%, 99.2%): dropping after 14
```

- add a cache: results in a bigger performance impact

[^1]: i've incremented a counter in the place I now decrement the depth at and then printed it on drop

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
if this ends up neutral then r=me

File: compiler/rustc_type_ir/src/fast_reject.rs [hunk 1]
```diff
@@ -273,6 +278,12 @@ impl<I: Interner, const INSTANTIATE_LHS_WITH_INFER: bool, const INSTANTIATE_RHS_
             | ty::Placeholder(_) => {}
         };
 
+        self.num_non_trivial_types += 1;
```
Comment (oli-obk) [inline compiler/rustc_type_ir/src/fast_reject.rs:281]: Maybe move this into `types_may_unify` and `args_may_unify` to only pay thr cost when recursing?

Comment (lcnr) [reply compiler/rustc_type_ir/src/fast_reject.rs:281]: the biggest one is typenum, tbh I would not be surprised if typenum actually relies on fast reject with large enough types

Comment (oli-obk) [reply compiler/rustc_type_ir/src/fast_reject.rs]: Oh... yea I think we can treat typenum "regressions below 100% are fine" if sufficiently motivated. Which this PR is.

**Comment by compiler-errors** [comment]
I think we need to eat this perf regression to make the trait solver feasible to land :>

r=me

**Comment by lqd** [comment]
Only a small regression, multiple strategies were tried here and in other PRs to get to this result — and it's a necessary hit.

> I think we need to eat this perf regression to make the trait solver feasible to land

from https://github.com/rust-lang/rust/pull/133566#issuecomment-2508717472

@rustbot label: +perf-regression-triaged

**Pull request merged**